### PR TITLE
Issue #1373 add test plan

### LINKED ISSUE #1373: Write test plan of things that need to be tested for 1.0 release
@deltamarnix mentioned that it is good to first write up what we want to test additionally to the stuff currently in our CI. We can then work on automating these if possible.

Test cases I'm currently thinking of:

- https://github.com/Deltares/imod-python/issues/988
- Tutorial material

Broad metrics we could test for:

- Did the test run without error, did the validation succeed? 
- Are the model input files written by iMOD Python the same (account for differences in sorting of List-based input)?
- Are iMOD Python's run times in a similar ballpark?
- Are there large differences in memory usage? Did memory usage explode?
- How much does model output differ? (Compute some basic statistics, mean, min, max difference)
- Make some plots of model output differences, these can help identifying differences.

### PR DESCRIPTION
**Pull request by JoerivanEngelen**
Fixes #1373

# Description
Adds test plan for the 1.0 release. Two remarks:

- The LHM stress test information is still aspecific on where to download data etc, as we are still conducting some tests here. We haven't fully arrived at a final test case here (we are still tweaking the case). When we have finished this work, we can upload it on a MinIO bucket.
- The pixi task to generate a grid to be imported for qgis still has to be made

# Checklist

- [x] Links to correct issue
- [ ] Update changelog, if changes affect users
- [x] PR title starts with ``Issue #nr``, e.g. ``Issue #737``
- [ ] Unit tests were added
- [ ] **If feature added**: Added/extended example

### CODE REVIEW

File: docs/developing/test-plan.rst [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+**This page is still under construction**.
+
+This document describes how and when to perform tests for iMOD Python.
+
+Known shortcomings and issues can be documented `here
+<https://deltares.github.io/imod-python/faq/known-issues.html>`_ . Bugs can be
+reported on `GitHub <https://github.com/Deltares/imod-python/issues>`_ .
+
+Functional tests
+----------------
```
Comment (deltamarnix) [inline docs/developing/test-plan.rst:13]: Under functional testing I would also expect some edge cases testers can think of to make things go wrong.

File: docs/developing/test-plan.rst [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+   directory of your liking.
+2. Assign the environment variable ``LHM_PRJ`` to the path of the projectfile
+   of the test data. You can put this in the ``.env`` file in the root of the
+   repository, or set it in your shell.
+3. Run the user acceptance tests by running the following command in the root 
+   of the repository:
+
+.. code-block:: console
+
+  pixi run user_acceptance
```
Comment (deltamarnix) [inline docs/developing/test-plan.rst:56]: Sounds like something we could automate in the future.

Step 1 would be to run the performance tests before every release (or more often), and make the output available for download so that a user can check it. This makes the machines more reliable, e.g. when we use a docker image.
Step 2 would be to also automate the checks that the human eye performs after the tests are run.

File: docs/developing/test-plan.rst [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+the repository:
+
+.. code-block:: console
+
+  pixi run docs
+
+Check if the documentation builds without errors and warnings. If there are
+errors or warnings, fix them before releasing in a pull request on `Github
+<https://github.com/Deltares/imod-python/pulls>`_ . Next, check if the
+documentation pages are rendered correctly and if the information on them is not
```
Comment (deltamarnix) [inline docs/developing/test-plan.rst:104]: This also sounds like something we can automate, I can imagine there are UI automation tests, maybe with Ranorex, that can test for HTML and CSS output.

**Review by deltamarnix — CHANGES_REQUESTED** [review]
This is just the start of the document. The setup is good enough, but details are lacking for a tester to be able to check off boxes while testing.

File: docs/developing/test-plan.rst [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+* The conversion of the transient LHM model run of 40 years on a daily timestep
+  (140K stress-periods) should be possible without memory overflow.
+
+Manual checks
+*************
+
+- Run the pixi task written below. This will export data to a UGRID NetCDF and
+  save it under the name <......>. Open QGIS. Click "Layers" > "Add Layer" >
+  "Add mesh". Insert the path <......> in the text box. This will import the
+  mesh. Verify if the mesh is rendered properly, if not open an issue on `GitHub
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply docs/developing/test-plan.rst]: I've added more explicit information what to look for to the vaguely worded things you pointed out.

File: docs/developing/test-plan.rst [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+*****************
+
+These are stress tests that tests the capabilities of iMOD Python in dealing
+with large groundwater models. The model currently used for the performance
+tests is `the LHM model <https://nhi.nu/modellen/lhm/>`_, but more models might
+be added in the future.
+
+Run the performance tests locally by following these steps:
+
+1. Download the test data from <...insert...link...here...> and put it in a
```
Comment (deltamarnix) [inline docs/developing/test-plan.rst:64]: Please insert the link

File: docs/developing/test-plan.rst [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+  pixi run user_acceptance
+
+Criteria for user acceptance tests of the 1.0 release are:
+
+* The tests should run without errors.
+* The tests should run without warnings from iMOD Python, unless unavoidable.
+* The MODFLOW6 and MetaSWAP input files written by iMOD Python should be the
+  same as iMOD5 (accounting for differences in row sorting.), unless there was a
+  conscious decision to divert from this.
```
Comment (deltamarnix) [inline docs/developing/test-plan.rst:82]: Where do you keep track of the conscious decisions?

File: docs/developing/test-plan.rst [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+* The conversion from projectfile to MODFLOW6 and MetaSWAP input files should be
+  done in a reasonable amount of time and should not be much slower than iMOD5.
```
Comment (deltamarnix) [inline docs/developing/test-plan.rst:84]: This is too vague. Add some percentages like:

```suggestion
  compared to the same conversion process of iMOD5. Differences in performance may not exceed 5%.
```

Comment (deltamarnix) [reply docs/developing/test-plan.rst:84]: Also, are the iMOD5 tests run for comparison on the same machine? Add instructions on how to do that.

File: docs/developing/test-plan.rst [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+* The conversion of the transient LHM model run of 40 years on a daily timestep
+  (140K stress-periods) should run without memory overflow on a machine with 32
+  GB and write a model within 30 minutes.
+
+Manual checks
+*************
+
+- Run the pixi task written below. This will export data to a UGRID NetCDF and
+  save it under the name <......>. Open QGIS. Click "Layers" > "Add Layer" >
+  "Add mesh". Insert the path <......> in the text box. This will import the
```
Comment (deltamarnix) [inline docs/developing/test-plan.rst:97]: Insert the paths please

File: docs/developing/test-plan.rst [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+  mesh. Verify if the mesh is rendered in two dimensions, and not as a single
+  line of cells. If not open an issue on `GitHub
```
Comment (deltamarnix) [inline docs/developing/test-plan.rst:99]:
```suggestion
  line of cells. If not, open an issue on `GitHub
```

File: docs/developing/test-plan.rst [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  <https://github.com/Deltares/imod-python/issues>`_ . 
+  
+.. code-block:: console
+
+  pixi run export-qgis
+
+- Work through the tutorial material here and verify it is up to date:
+  https://deltares.github.io/iMOD-Documentation/ . If it is not, open an issue
```
Comment (deltamarnix) [inline docs/developing/test-plan.rst:107]: How do we verify if it is up to date? Add instructions or refer to requirements.

File: docs/developing/test-plan.rst (hunk 3 above)
Comment (deltamarnix) [reply docs/developing/test-plan.rst:124]: If you could create an issue, you can close this comment.

**Review by deltamarnix — CHANGES_REQUESTED** [review]
There are still some missing links and instructions.

File: docs/developing/test-plan.rst (hunk 7 above)
Comment (JoerivanEngelen) [reply docs/developing/test-plan.rst:110]: Added both.

File: docs/developing/test-plan.rst (hunk 10 above)
Comment (JoerivanEngelen) [reply docs/developing/test-plan.rst]: Changed to:

```
1. `Open the tutorial material here
   <https://deltares.github.io/iMOD-Documentation/tutorial_Hondsrug.html/>`_ .
2. Run each jupyter notebook and assure it runs without errors.
3. If there are any errors, open an issue on `iMOD Documentation repository
   Github <https://github.com/Deltares/iMOD-Documentation/issues>`_ .
```

File: docs/developing/test-plan.rst [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+- Installing iMOD Python and running tests in a folder with curly braces ``{}``
+  in the name, like is unavoidable on the Windows Computational Facilities
+  (WCF). 
+- Installing iMOD Python with the iMODforge distribution on a machine
+  disconnected from the internet, like Deltares' LHM server. Examples that use
+  the ``imod.data`` module here will not work, as these require an internet
+  connection to download the data.
+- Installing iMOD Python on a machine that runs on a different operating system
+  than Windows, than which is what iMOD Python is mostly developend and used on.
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply docs/developing/test-plan.rst]: Changed to:

```
- Installing iMOD Python on Linux. This is less commonly used by iMOD Python
  users and developers, therefore might lead to missed edge cases.
```

File: docs/developing/test-plan.rst (hunk 3 above)
Comment (JoerivanEngelen) [reply docs/developing/test-plan.rst:166]: Done https://github.com/Deltares/imod-python/issues/1615

**Comment by JoerivanEngelen** [comment]
UPDATE:

The user acceptance data (LHM stress test) job is now finished. I versioned the data using DVC, so that changes to the data can be committed to the git repo, without having to push the actual data to Github. Instead, it's pushed/pulled from a Deltares MinIO storage. I updated the instructions accordingly, increasing clarity a lot.
